src/QsCompiler/Transformations/ClassicallyControlled.cs

```diff
@@ -870,7 +929,7 @@ public override QsStatementKind OnConditionalStatement(QsConditionalStatement st
                         }
 
                         this.SharedState.GeneratedOpParams = contextParams;
-                        this.SharedState.IsValidScope = contextValidScope;
+                        this.SharedState.IsValidScope = this.SharedState.IsValidScope && contextValidScope;
```

Comment:
I am thinking that it may be the case where an inner scope is not valid for lifting, but an outer scope may be.
For example if there is a mutable variable defined in the outer scope that is updated in the inner scope. The inner scope would not be valid because it is setting a mutable variable defined outside of itself, but the outer scope would be fine for lifting because that mutable is defined inside itself.

Reply:
Oh, that's a good point. I had been so focused on the scenario where the mutable is defined outside of the outer scope and then modified inside the inner scope. This may not be as easy of a fix then, because an inner scope that is not valid for lifting would need to return which variables made it invalid for lifting so that the outer scope could then differentiate between those mutable updates that were made to something declaring in the outer scope vs not. I'll follow up with you to get your thoughts on how to handle this, but it may be that this is better saved for the larger review and fix to the use of `IsValidScope` that is pending.

Reply:
It should work the following way:
```
operation Foo() {
    if (...) {
        mutable x = 3;
        if (...) {
            set x <- 4;
        }
    }
}
```
The inner scope is evaluated first and determined to be invalid because of the mutable set. The state after this is:
```
operation Foo() {
    if (...) {
        mutable x = 3;
        if (...) {
            set x <- 4;
        }
    }
}
```
Then the outer scope is evaluated and looks at its entire contents at that point. It is determined to be valid for lifting because the mutable variable that is set is defined in its scope. The state after this is:
```
LiftedOp (...) {
    mutable x = 3;
    if (...) {
        set x <- 4;
    }
}
...
operation Foo() {
    LiftedOp(...);
}
```

Reply:
If instead we have an invalidation due to return:
```
operation Foo() {
    if (...) {
        if (...) {
            return;
        }
    }
}
```
Then the inner scope is determined to be invalid from having the return.
Then the outer scope is independently evaluated and seen to be invalid from the same return.
This results in no lifting and no change.

Reply:
If we have:
```
operation Foo() {
    if (...) {
        if (...) {
            // Valid code for lifting
        }
        return;
    }
}
```
Then the inner scope is valid, resulting in:
```
LiftedOp (...) {
    // Valid code for lifting
}
...
operation Foo() {
    if (...) {
        LiftedOp(...);
        return;
    }
}
```
Then the outer scope is evaluated and found to be invalid.

Reply:
Finally, a fully valid case:
```
operation Foo() {
    if (...) {
        if (...) {
            // Valid for lifting
        }
        // Also valid for lifting
    }
}
```
Then the inner scope is evaluated, found valid, and lifted:
```
operation LiftOp1(...) {
    // Valid for lifting
}
...
operation Foo() {
    if (...) {
        LiftOp1(...);
        // Also valid for lifting
    }
}
```
Then the outer scope is evaluated, found valid, and lifted:
```
operation LiftOp1(...) {
    // Valid for lifting
}
...
operation LiftOp2(...) {
    LiftOp1(...);
    // Also valid for lifting
}
...
operation Foo() {
    LiftOp2(...);
}
```

Reply:
Yeah, that makes sense.  The case I'm worried about is this:
```qsharp
operation Foo() {
    mutable x = 3;
    if (...) {
        if (...) {
            set x = 4;
        }
    }
}
```
In that case, the inner scope correctly doesn't get lifted, but then the outer scope DOES get lifted because it didn't have an explicit set inside of it (the nested set isn't seen). My approach of just propagating up the validity of the inner scope fixes this bug but breaks the example you posted. We need some way to differentiate inner scope invalidity that should propagate to the parent and those that shouldn't, which would depend on where the initial definition of the mutable variable came from.

Reply:
Yah after testing this I can see that it doesn't do what is expected. The whole scope needs to be evaluated again when it comes to the outer scope, but maybe that is not efficient? Either way this seems like an open issue.

Reply:
Yeah, agreed. Since it just trades one issue for another, I'll remove the change I made for this here and file it as a separate issue.

Reply:
Filed as https://github.com/microsoft/qsharp-compiler/issues/768

Reply:
Thanks for filing that issue: #768